waku/v2/node/builder.nim

```diff
@@ -0,0 +1,167 @@
+    switchSslSecureCert: Option[string]
+    switchSendSignedPeerRecord: Option[bool]
+
+    # Waku discv5
+    wakuDiscv5: Option[WakuDiscoveryV5]
+
+  WakuNodeBuilderResult* = Result[void, string]
+
+
+## Init
+
+proc init*(T: type WakuNodeBuilder): WakuNodeBuilder =
+  WakuNodeBuilder()
+
+
+## General
+
+proc withRng*(builder: var WakuNodeBuilder, rng: ref crypto.HmacDrbgContext) =
+  builder.nodeRng = some(rng)
+
+proc withNodeKey*(builder: var WakuNodeBuilder, nodeKey: crypto.PrivateKey) =
+  builder.nodeKey = some(nodeKey)
+
+proc withNetworkConfiguration*(builder: var WakuNodeBuilder, config: NetConfig) =
+  builder.netConfig = some(config)
+
+proc withNetworkConfigurationDetails*(builder: var WakuNodeBuilder,
+          bindIp: ValidIpAddress,
+          bindPort: Port,
+          extIp = none(ValidIpAddress),
+          extPort = none(Port),
+          extMultiAddrs = newSeq[MultiAddress](),
+          wsBindPort: Port = Port(8000),
+          wsEnabled: bool = false,
+          wssEnabled: bool = false,
+          wakuFlags = none(CapabilitiesBitfield),
+          dns4DomainName = none(string),
+          discv5UdpPort = none(Port)): WakuNodeBuilderResult {.
+  deprecated: "use 'builder.withNetworkConfiguration()' instead".} =
+  let netConfig = ? NetConfig.init(
+    bindIp = bindIp,
+    bindPort = bindPort,
+    extIp = extIp,
+    extPort = extPort,
+    extMultiAddrs = extMultiAddrs,
+    wsBindPort = wsBindPort,
+    wsEnabled = wsEnabled,
+    wssEnabled = wssEnabled,
+    wakuFlags = wakuFlags,
+    dns4DomainName = dns4DomainName,
+    discv5UdpPort = discv5UdpPort,
+  )
+  builder.withNetworkConfiguration(netConfig)
+  ok()
+
+
+## Peer storage and peer manager
+
+proc withPeerStorage*(builder: var WakuNodeBuilder, peerStorage: PeerStorage, capacity = none(int)) =
+  if not peerStorage.isNil():
+    builder.peerStorage = some(peerStorage)
+
+  builder.peerStorageCapacity = capacity
+
+proc withPeerManager*(builder: var WakuNodeBuilder, peerManager: PeerManager) =
+  builder.peerManager = some(peerManager)
+
+
+## Waku switch
+
+proc withSwitchConfiguration*(builder: var WakuNodeBuilder,
+                              maxConnections = none(int),
+                              nameResolver: NameResolver = nil,
+                              sendSignedPeerRecord = false,
+                              secureKey = none(string),
+                              secureCert = none(string),
+                              agentString = none(string)) =
+  builder.switchMaxConnections = maxConnections
+  builder.switchSendSignedPeerRecord = some(sendSignedPeerRecord)
+  builder.switchSslSecureKey = secureKey
+  builder.switchSslSecureCert = secureCert
+  builder.switchAgentString = agentString
+
+  if not nameResolver.isNil():
+    builder.switchNameResolver = some(nameResolver)
+
+
+## Waku discv5
+
+proc withWakuDiscv5*(builder: var WakuNodeBuilder, instance: WakuDiscoveryV5) =
+  if not instance.isNil():
```

Comment:
Should we do anything for the case 'isNil' ?

Reply:
I added this for compatibility reasons (with the deprecated `WakuNode.new()` API). Accepting a `nil` makes 8 test cases crash 😅